src/apprt/gtk/Surface.zig

```diff
@@ -847,28 +847,41 @@ pub fn shouldClose(self: *const Surface) bool {
 }
 
 pub fn getContentScale(self: *const Surface) !apprt.ContentScale {
-    // Future: detect GTK version 4.12+ and use gdk_surface_get_scale so we
-    // can support fractional scaling.
-    const gtk_scale: f32 = @floatFromInt(c.gtk_widget_get_scale_factor(@ptrCast(self.gl_area)));
+    const gtk_scale: f32 = scale: {
+        const widget: *gtk.Widget = @ptrCast(@alignCast(self.gl_area));
+        // Future: detect GTK version 4.12+ and use gdk_surface_get_scale so we
+        // can support fractional scaling.
+        const scale = widget.getScaleFactor();
+        if (scale < 0) {
+            log.warn("gtk_widget_get_scale_factor returned a negative number: {d:.3}", .{scale});
+            break :scale 1.0;
+        }
+        break :scale @floatFromInt(scale);
+    };
 
     // Also scale using font-specific DPI, which is often exposed to the user
     // via DE accessibility settings (see https://docs.gtk.org/gtk4/class.Settings.html).
     const xft_dpi_scale = xft_scale: {
         // gtk-xft-dpi is font DPI multiplied by 1024. See
         // https://docs.gtk.org/gtk4/property.Settings.gtk-xft-dpi.html
-        const settings = c.gtk_settings_get_default();
-
-        var value: c.GValue = std.mem.zeroes(c.GValue);
-        defer c.g_value_unset(&value);
-        _ = c.g_value_init(&value, c.G_TYPE_INT);
-        c.g_object_get_property(@ptrCast(@alignCast(settings)), "gtk-xft-dpi", &value);
-        const gtk_xft_dpi = c.g_value_get_int(&value);
+        const settings = gtk.Settings.getDefault() orelse break :xft_scale 1.0;
+        var value = std.mem.zeroes(gobject.Value);
+        defer value.unset();
+        _ = value.init(gobject.ext.typeFor(c_int));
+        settings.as(gobject.Object).getProperty("gtk-xft-dpi", &value);
+        const gtk_xft_dpi = value.getInt();
+
+        // ensure that the scale is never negative
+        if (gtk_xft_dpi < 0) {
```

Comment:
Here's a crazy idea... what if the negative content scale problem was caused only by gtk-xft-dpi being negative? [The docs](https://docs.gtk.org/gtk4/property.Settings.gtk-xft-dpi.html) say apparently it can be -1 to signify the "default" value (whatever that means). We should corroborate this with users, obviously, but I suspect this is what's happening.

Reply:
That begs the question of what the "default" value is. From what I've googled it seems as if the default is whatever works out to a scale factor of 1.0, so basically no code changes necessary. Would be good to add that as a default.

Reply:
These commits to libadwaita seems to confirm `gtk-xft-dpi` <= 0 should be treated as 96 * 1024:

https://gitlab.gnome.org/GNOME/libadwaita/-/commit/a7738a4d269bfdf4d8d5429ca73ccdd9b2450421
https://gitlab.gnome.org/GNOME/libadwaita/-/merge_requests/1372/diffs?commit_id=9759d3fd81129608dd78116001928f2aed974ead